imperial_coldfront_plugin/microsoft_graph_client.py

```diff
@@ -0,0 +1,78 @@
+"""Interface for interacting with the Microsoft Graph API."""
+
+import requests
+from django.conf import settings
+from uplink import Consumer, get, response_handler
+
+
+def _get_app_access_token():
+    """Get an access token for the application to use the Microsoft Graph API.
+
+    Fetches an access token that is enabled for app-only access i.e. not on behalf of a
+    logged in user.
+    """
+    tenant_id = settings.MICROSOFT_TENANT_ID
+    response = requests.post(
+        f"https://login.microsoftonline.com/{tenant_id}/oauth2/v2.0/token",
+        data={
+            "grant_type": "client_credentials",
+            "client_id": settings.OIDC_RP_CLIENT_ID,
+            "client_secret": settings.OIDC_RP_CLIENT_SECRET,
+            "scope": "https://graph.microsoft.com/.default",
+        },
+    )
+    return response.json()["access_token"]
+
+
+def parse_profile_data(response):
+    """Parse the user profile data from the API response into a useful format."""
+    data = response.json()
+    extension_attributes = data.get("onPremisesExtensionAttributes", {})
+    return {
+        "job_title": data.get("jobTitle"),
+        "department": data.get("department"),
+        "company_name": data.get("companyName"),
+        "user_type": data.get("userType"),
+        "job_family": extension_attributes.get("extensionAttribute14"),
+        "employment_status": extension_attributes.get("extensionAttribute6"),
+    }
+
+
+def get_uid_from_response(response):
+    """Extract the Unix uid from the API response."""
+    data = response.json()
+    uid = data.get("onPremisesExtensionAttributes", {}).get("extensionAttribute12")
+    return uid if uid is None else int(uid)
+
+
+class MicrosoftGraphClient(Consumer):
+    """Client for interacting with the Microsoft Graph API.
+
+    Provides an abstracted interface for making requests to the Microsoft Graph API that
+    provides easy to use data structures for the responses.
+    """
+
+    @response_handler(parse_profile_data)
+    @get(
+        "users/{username}@ic.ac.uk?$select=jobTitle,department,companyName,userType,onPremisesExtensionAttributes"
+    )
+    def user_profile(self, username: str):
+        """Get the profile data for a user."""
+        pass
+
+    @response_handler(get_uid_from_response)
+    @get("users/{username}@ic.ac.uk?$select=onPremisesExtensionAttributes")
+    def user_uid(self, username: str):
+        """Get the Unix uid for a user."""
```

Comment:
On that note, what normally goes in the function with uplink, if all is handled in the handler?

Reply:
Nothing. The method is sort of just a place to hang decorators off of.